javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
-              // precision here means number of significant digits
-              // not digits after decimal point
-              val p =
-                if (!hasPrecision) 6
-                else if (precision == 0) 1
-                else precision
-              // between 1e-4 and 10e(p): display as fixed
-              if (m >= 1e-4 && m < Math.pow(10, p)) {
-                /* First approximation of the smallest power of 10 that is >= m.
-                 * Due to rounding errors in the event of an imprecise `log10`
-                 * function, sig0 could actually be the smallest power of 10
-                 * that is > m.
-                 */
-                val sig0 = Math.ceil(Math.log10(m)).toInt
-                /* Increment sig0 so that it is always the first power of 10
-                 * that is > m.
-                 */
-                val sig = if (Math.pow(10, sig0) <= m) sig0 + 1 else sig0
-                with_+(numberArg.toFixed(Math.max(p - sig, 0)))
-              } else sciNotation(p - 1)
-            case 'f' =>
-              with_+({
-                // JavaDoc: 6 is default precision
-                numberArg.toFixed(if (hasPrecision) precision else 6)
-              }, numberArg.isNaN || numberArg.isInfinite)
-          }
+            formatNonNumericString(flags, width, -1, str.asInstanceOf[String])
+          case _ =>
+            formatNullOrThrowIllegalFormatConversion()
+        }
+
+      case 'd' =>
+        validateFlags(flags, conversion, AltFormat)
+        rejectPrecision()
+        arg match {
+          case arg: Int =>
+            formatNumericString(flags, width, arg.toString())
+          case arg: Long =>
+            formatNumericString(flags, width, arg.toString())
+          case _ =>
+            formatNullOrThrowIllegalFormatConversion()
+        }
+
+      case 'o' =>
+        validateFlags(flags, conversion, InvalidFlagsForOctalAndHex)
+        rejectPrecision()
+        val prefix =
+          if (flags.altFormat) "0"
+          else ""
+        arg match {
+          case arg: Int =>
+            padAndSendToDest(flags, width, prefix,
+                java.lang.Integer.toOctalString(arg))
+          case arg: Long =>
+            padAndSendToDest(flags, width, prefix,
+                java.lang.Long.toOctalString(arg))
+          case _ =>
+            formatNullOrThrowIllegalFormatConversion()
+        }
+
+      case 'x' | 'X' =>
+        validateFlags(flags, conversion, InvalidFlagsForOctalAndHex)
+        rejectPrecision()
+        val prefix = {
+          if (!flags.altFormat) ""
+          else if (flags.upperCase) "0X"
+          else "0x"
+        }
+        arg match {
+          case arg: Int =>
+            padAndSendToDest(flags, width, prefix,
+                applyUpperCase(flags, java.lang.Integer.toHexString(arg)))
+          case arg: Long =>
+            padAndSendToDest(flags, width, prefix,
+                applyUpperCase(flags, java.lang.Long.toHexString(arg)))
```

Comment:
Seems like it might be worth introducing a `prefix` parameter to `formatNonNumericString`?

Reply:
I considered that, but it doesn't fit. `prefix` is the thing that goes after leading space padding, but before 0 padding. If I introduce a `prefix` in `formatNonNumericString`, then I need to take 0 padding into account in that method, and that complicates things. It was cleaner in the end to only repeat `applyUpperCase` in the `x`/`X` conversion.